### PR TITLE
r/aws_sso_permission_set: New Resource

### PR DESCRIPTION
**Pull request by lawdhavmercy**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates #15108
Relates #15540

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
data-source/aws_ssoadmin_instances: new data-source
data-source/aws_ssoadmin_permission_set: new data-source
resource/aws_ssoadmin_permission_set: new resource
resource/aws_ssoadmin_permission_set_inline_policy: new resource
resource/aws_ssoadmin_managed_policy_attachment: new resource
```

Output from acceptance testing:

Commerical Organizations-Only
```
--- PASS: TestAccAWSProvider_AssumeRole_Empty (57.99s)
--- PASS: TestAccAWSProvider_Endpoints (54.60s)
--- PASS: TestAccAWSProvider_IgnoreTags_EmptyConfigurationBlock (53.59s)
--- PASS: TestAccAWSProvider_IgnoreTags_KeyPrefixes_Multiple (51.08s)
--- PASS: TestAccAWSProvider_IgnoreTags_KeyPrefixes_None (52.84s)
--- PASS: TestAccAWSProvider_IgnoreTags_KeyPrefixes_One (53.78s)
--- PASS: TestAccAWSProvider_IgnoreTags_Keys_Multiple (52.68s)
--- PASS: TestAccAWSProvider_IgnoreTags_Keys_None (51.76s)
--- PASS: TestAccAWSProvider_IgnoreTags_Keys_One (51.93s)
--- PASS: TestAccAWSProvider_Region_AwsChina (39.31s)
--- PASS: TestAccAWSProvider_Region_AwsCommercial (39.64s)
--- PASS: TestAccAWSProvider_Region_AwsGovCloudUs (39.68s)
--- PASS: TestAccAWSSSOAdminManagedPolicyAttachment_basic (68.53s)
--- PASS: TestAccAWSSSOAdminManagedPolicyAttachment_disappears (63.25s)
--- PASS: TestAccAWSSSOAdminManagedPolicyAttachment_disappears_permissionSet (62.04s)
--- PASS: TestAccAWSSSOAdminManagedPolicyAttachment_forceNew (121.07s)
--- PASS: TestAccAWSSSOAdminManagedPolicyAttachment_multipleManagedPolicies (115.87s)
--- PASS: TestAccAWSSSOAdminPermissionSetInlinePolicy_basic (59.10s)
--- PASS: TestAccAWSSSOAdminPermissionSetInlinePolicy_disappears (51.81s)
--- PASS: TestAccAWSSSOAdminPermissionSetInlinePolicy_disappears_permissionSet (53.72s)
--- PASS: TestAccAWSSSOAdminPermissionSetInlinePolicy_update (98.95s)
--- PASS: TestAccAWSSSOAdminPermissionSet_basic (46.09s)
--- PASS: TestAccAWSSSOAdminPermissionSet_mixedPolicyAttachments (85.43s)
--- PASS: TestAccAWSSSOAdminPermissionSet_tags (105.08s)
--- PASS: TestAccAWSSSOAdminPermissionSet_updateDescription (85.98s)
--- PASS: TestAccAWSSSOAdminPermissionSet_updateRelayState (85.58s)
--- PASS: TestAccAWSSSOAdminPermissionSet_updateSessionDuration (85.20s)
--- PASS: TestAccDataSourceAWSSSOAdminInstances_basic (55.83s)
--- PASS: TestAccDataSourceAWSSSOAdminPermissionSet_arn (58.78s)
--- PASS: TestAccDataSourceAWSSSOAdminPermissionSet_name (58.74s)
--- PASS: TestAccDataSourceAWSSSOAdminPermissionSet_nonExistent (11.14s)
--- PASS: TestProvider (1.04s)
--- PASS: TestProvider_impl (0.31s)
```
Commerical:
```
--- PASS: TestAccAWSProvider_AssumeRole_Empty (32.46s)
--- PASS: TestAccAWSProvider_Endpoints (31.21s)
--- PASS: TestAccAWSProvider_IgnoreTags_EmptyConfigurationBlock (30.26s)
--- PASS: TestAccAWSProvider_IgnoreTags_KeyPrefixes_Multiple (31.47s)
--- PASS: TestAccAWSProvider_IgnoreTags_KeyPrefixes_None (31.60s)
--- PASS: TestAccAWSProvider_IgnoreTags_KeyPrefixes_One (30.97s)
--- PASS: TestAccAWSProvider_IgnoreTags_Keys_Multiple (29.94s)
--- PASS: TestAccAWSProvider_IgnoreTags_Keys_None (31.66s)
--- PASS: TestAccAWSProvider_IgnoreTags_Keys_One (31.14s)
--- PASS: TestAccAWSProvider_Region_AwsChina (25.88s)
--- PASS: TestAccAWSProvider_Region_AwsCommercial (26.63s)
--- PASS: TestAccAWSProvider_Region_AwsGovCloudUs (25.92s)
--- PASS: TestProvider (0.53s)
--- PASS: TestProvider_impl (0.27s)
--- SKIP: TestAccAWSSSOAdminManagedPolicyAttachment_basic (0.89s)
--- SKIP: TestAccAWSSSOAdminManagedPolicyAttachment_disappears (0.76s)
--- SKIP: TestAccAWSSSOAdminManagedPolicyAttachment_disappears_permissionSet (0.77s)
--- SKIP: TestAccAWSSSOAdminManagedPolicyAttachment_forceNew (1.11s)
--- SKIP: TestAccAWSSSOAdminManagedPolicyAttachment_multipleManagedPolicies (1.23s)
--- SKIP: TestAccAWSSSOAdminPermissionSetInlinePolicy_basic (1.04s)
--- SKIP: TestAccAWSSSOAdminPermissionSetInlinePolicy_disappears (1.08s)
--- SKIP: TestAccAWSSSOAdminPermissionSetInlinePolicy_disappears_permissionSet (0.88s)
--- SKIP: TestAccAWSSSOAdminPermissionSetInlinePolicy_update (1.14s)
--- SKIP: TestAccAWSSSOAdminPermissionSet_basic (1.10s)
--- SKIP: TestAccAWSSSOAdminPermissionSet_mixedPolicyAttachments (1.05s)
--- SKIP: TestAccAWSSSOAdminPermissionSet_tags (1.25s)
--- SKIP: TestAccAWSSSOAdminPermissionSet_updateDescription (1.06s)
--- SKIP: TestAccAWSSSOAdminPermissionSet_updateRelayState (0.88s)
--- SKIP: TestAccAWSSSOAdminPermissionSet_updateSessionDuration (0.96s)
--- SKIP: TestAccDataSourceAWSSSOAdminInstances_basic (1.12s)
--- SKIP: TestAccDataSourceAWSSSOAdminPermissionSet_arn (1.17s)
--- SKIP: TestAccDataSourceAWSSSOAdminPermissionSet_name (1.08s)
--- SKIP: TestAccDataSourceAWSSSOAdminPermissionSet_nonExistent (0.95s)
```

### CODE REVIEW

File: aws/data_source_aws_sso_permission_set.go
Comment (lachlancooper) [inline aws/data_source_aws_sso_permission_set.go:163]:
```suggestion
		return fmt.Errorf("Error listing tags for AWS SSO Permission Set (%s): %s", permissionSetArn, tagsErr)
```

**Comment by anGie44** [comment]
Hi @lawdhavmercy @burck1, thank you for this PR and your patience! Are you still interested and available to work on this? We have been unable to get to the SSO service until now but we aim to get this work in before the end of the year. 

To be upfront about the process, here is what to expect: A design for this service, essentially covering the resources/data-sources we feel are essential for a practitioner's enablement, has been created internally and is currently in review by our team.  Once the design is approved, implementation will need to follow the design. Hopefully, we can use some or all of your work to support this core service. We won't know until we have concensus on the design, however. Once we have approval, I'll circle back here and provide an update/initial code review.

**Comment by burck1** [comment]
Hi @anGie44! Great to hear. Once the design is finalized, we'd be happy to help migrate our code to the updated design. Feel free to reach out as soon as the design is ready.

One nuance of the ssoadmin apis that may help inform your design is the [ProvisionPermissionSet](https://docs.aws.amazon.com/singlesignon/latest/APIReference/API_ProvisionPermissionSet.html) request. This API must be called after any updates to the permission set resource or else the change will not actually be propagated to the assigned accounts. This is in fact the reason why we decided not to create separate aws_ssoadmin_permission_set_inline_policy and aws_ssoadmin_managed_policy_attachment resources and instead made those properties of the permission set resource. I.E. Whenever those policies are modified, the ProvisionPermissionSet API must be called. Separating out those resources means you would need to call the ProvisionPermissionSet API within the update function for each of those resources which means you would run into conflicts whenever more than one of those resources has changed. The API prevents multiple provisionings to the same assigned account from happening at once. The solution we ended up coming up with was simply to incorporate the inline and managed policies as properties of the permission set resource so that we would only need to call the ProvisionPermissionSet API one time within the permission set resource's update function. I discussed this a bit in this comment: https://github.com/hashicorp/terraform-provider-aws/pull/15322#issuecomment-705672143

Let us know if you have any questions.

Note: We have been running our custom version of the aws provider with the changes from #15322 in production for over a month now without any problems.

**Comment by burck1** [comment]
The code for calling the ProvisionPermissionSet API can be found here: https://github.com/hashicorp/terraform-provider-aws/blob/1b750b9ce4e519cfff995df4bbf309d956d53a0e/aws/resource_aws_sso_permission_set.go#L483

**Comment by anGie44** [comment]
Thanks for clarifying @burck1  -- definitely to our advantage that these changes have been tested in production 👍  My only hesitation in not moving out the policy related attributes to their own resources is the need to make various API calls to gather these attributes on import/read which strays a bit from convention. You mentioned in your testing you came across the need to call `ProvisionPermissionSet` on all types of resource updates, though in the API docs, AWS only explicitly notes to use `ProvisionPermissionSet` after use of `AttachManagedPolicyToPermissionSet` and `PutInlinePolicyToPermissionSet` and interestingly not on `UpdatePermissionSet` 🤔 -- just wanted to clarify if you confirmed it's indeed necessary to call `ProvisionPersimissionSet` on _all_ sort of updates? (definitely could be a miss though in the AWS documentation)

**Comment by burck1** [comment]
Hi @anGie44. Some great questions!

> My only hesitation in not moving out the policy related attributes to their own resources is the need to make various API calls to gather these attributes on import/read which strays a bit from convention.

True that is strays from convention. Unfortunately we couldn't find any other AWS resources in this repo that we could model our changes after that also had to follow this pattern of calling a different API (like `ProvisionPermissionSet`) after any changes.

> You mentioned in your testing you came across the need to call `ProvisionPermissionSet` on all types of resource updates, though in the API docs, AWS only explicitly notes to use `ProvisionPermissionSet` after use of `AttachManagedPolicyToPermissionSet` and `PutInlinePolicyToPermissionSet` and interestingly not on `UpdatePermissionSet` 🤔 -- just wanted to clarify if you confirmed it's indeed necessary to call `ProvisionPersimissionSet` on _all_ sort of updates? (definitely could be a miss though in the AWS documentation)

I think this is a miss in the AWS documentation. I just tested this out with the AWS CLI (IDs are obfuscated). Currently the only way to tell if you need to call the `ProvisionPersimissionSet` API for a given account assignment is to call the [ListAccountsForProvisionedPermissionSet](https://docs.aws.amazon.com/singlesignon/latest/APIReference/API_ListAccountsForProvisionedPermissionSet.html) API with the `ProvisioningStatus` parameter set to `LATEST_PERMISSION_SET_NOT_PROVISIONED`. This tells you that the permission set has been updated, but the change has not been pushed to assigned account(s). I've included this call in the CLI output.

```sh
# This is the current permission set settings
$ aws sso-admin describe-permission-set --instance-arn "arn:aws:sso:::instance/ssoins-9999999999999999" --permission-set-arn "arn:aws:sso:::permissionSet/ssoins-9999999999999999/ps-1111111111111111"
{
    "PermissionSet": {
        "Name": "alex-test",
        "PermissionSetArn": "arn:aws:sso:::permissionSet/ssoins-9999999999999999/ps-1111111111111111",
        "Description": "before update",
        "CreatedDate": "2020-12-11T17:18:37.654000-06:00",
        "SessionDuration": "PT1H"
    }
}

# There is currently one account assigned to the permission set that has been provisioned with the latest changes
$ aws sso-admin list-accounts-for-provisioned-permission-set --instance-arn "arn:aws:sso:::instance/ssoins-9999999999999999" --permission-set-arn "arn:aws:sso:::permissionSet/ssoins-9999999999999999/ps-1111111111111111" --provisioning-status LATEST_PERMISSION_SET_PROVISIONED
{
    "AccountIds": [
        "012345678910"
    ]
}

# There is currently no accounts assigned to the permission set that have not been provisioned with the latest changes
$ aws sso-admin list-accounts-for-provisioned-permission-set --instance-arn "arn:aws:sso:::instance/ssoins-9999999999999999" --permission-set-arn "arn:aws:sso:::permissionSet/ssoins-9999999999999999/ps-1111111111111111" --provisioning-status LATEST_PERMISSION_SET_NOT_PROVISIONED
{
    "AccountIds": []
}

# Let's update the description of the permission set
$ aws sso-admin update-permission-set --instance-arn "arn:aws:sso:::instance/ssoins-9999999999999999" --permission-set-arn "arn:aws:sso:::permissionSet/ssoins-9999999999999999/ps-1111111111111111" --description "after update"

# We can see the updated description
$ aws sso-admin describe-permission-set --instance-arn "arn:aws:sso:::instance/ssoins-9999999999999999" --permission-set-arn "arn:aws:sso:::permissionSet/ssoins-9999999999999999/ps-1111111111111111"
{
    "PermissionSet": {
        "Name": "alex-test",
        "PermissionSetArn": "arn:aws:sso:::permissionSet/ssoins-9999999999999999/ps-1111111111111111",
        "Description": "after update",
        "CreatedDate": "2020-12-11T17:18:37.654000-06:00",
        "SessionDuration": "PT1H"
    }
}

# but now there are currently no accounts assigned to the permission set that have been provisioned with the latest changes
$ aws sso-admin list-accounts-for-provisioned-permission-set --instance-arn "arn:aws:sso:::instance/ssoins-9999999999999999" --permission-set-arn "arn:aws:sso:::permissionSet/ssoins-9999999999999999/ps-1111111111111111" --provisioning-status LATEST_PERMISSION_SET_PROVISIONED    
{
    "AccountIds": []
}

# and our account now shows up as not being provisioned with the latest changes
$ aws sso-admin list-accounts-for-provisioned-permission-set --instance-arn "arn:aws:sso:::instance/ssoins-9999999999999999" --permission-set-arn "arn:aws:sso:::permissionSet/ssoins-9999999999999999/ps-1111111111111111" --provisioning-status LATEST_PERMISSION_SET_NOT_PROVISIONED
{
    "AccountIds": [
        "012345678910"
    ]
}

# so we can now call provision-permission-set
$ aws sso-admin provision-permission-set --instance-arn "arn:aws:sso:::instance/ssoins-9999999999999999" --permission-set-arn "arn:aws:sso:::permissionSet/ssoins-9999999999999999/ps-1111111111111111" --target-type ALL_PROVISIONED_ACCOUNTS
{
    "PermissionSetProvisioningStatus": {
        "Status": "IN_PROGRESS",
        "RequestId": "47ac8a85-2212-41d6-8bde-c69a7a81c834",
        "PermissionSetArn": "arn:aws:sso:::permissionSet/ssoins-9999999999999999/ps-1111111111111111"
    }
}

# and await for status to be SUCCEEDED
$ aws sso-admin describe-permission-set-provisioning-status --instance-arn "arn:aws:sso:::instance/ssoins-9999999999999999" --provision-permission-set-request-id "47ac8a85-2212-41d6-8bde-c69a7a81c834"
{
    "PermissionSetProvisioningStatus": {
        "Status": "SUCCEEDED",
        "RequestId": "47ac8a85-2212-41d6-8bde-c69a7a81c834",
        "PermissionSetArn": "arn:aws:sso:::permissionSet/ssoins-9999999999999999/ps-1111111111111111",
        "CreatedDate": "2020-12-11T17:50:26.049000-06:00"
    }
}

# and now our account is back to having the latest permission set changes provisioned
$ aws sso-admin list-accounts-for-provisioned-permission-set --instance-arn "arn:aws:sso:::instance/ssoins-9999999999999999" --permission-set-arn "arn:aws:sso:::permissionSet/ssoins-9999999999999999/ps-1111111111111111" --provisioning-status LATEST_PERMISSION_SET_PROVISIONED
{
    "AccountIds": [
        "012345678910"
    ]
}

# and none that are not at the latest
$ aws sso-admin list-accounts-for-provisioned-permission-set --instance-arn "arn:aws:sso:::instance/ssoins-9999999999999999" --permission-set-arn "arn:aws:sso:::permissionSet/ssoins-9999999999999999/ps-1111111111111111" --provisioning-status LATEST_PERMISSION_SET_NOT_PROVISIONED
{
    "AccountIds": []
}
```

Note: We also need to call `ProvisionPersimissionSet` after any calls to [TagResource](https://docs.aws.amazon.com/singlesignon/latest/APIReference/API_TagResource.html) or [UntagResource](https://docs.aws.amazon.com/singlesignon/latest/APIReference/API_UntagResource.html), even though that is not documented either.

I can also provide some screenshots of the AWS console as well if that helps.

**Comment by burck1** [comment]
> Note: to keep inline with the service package -> resource/data-source naming, we'll be using the **ssoadmin** prefix and **identitystore** where applicable

Regarding this note, does it make sense for @lawdhavmercy and I to go ahead and update this PR and #15322 as follows?

### Resources

* aws\_**sso**\_permission\_set → aws\_**ssoadmin**\_permission\_set
* aws\_**sso**\_assignment → aws\_**ssoadmin**\_assignment

### Data Sources

* aws\_**sso**\_instance → aws\_**ssoadmin**\_instance
* aws\_**sso**\_permission\_set → aws\_**ssoadmin**\_permission\_set
* aws\_**identity\_store**\_group → aws\_**identitystore**\_group
* aws\_**identity\_store**\_user → aws\_**identitystore**\_user

**Comment by burck1** [comment]
Also, regarding the `aws_ssoadmin_instance` data source, although the [ListInstances](https://docs.aws.amazon.com/singlesignon/latest/APIReference/API_ListInstances.html) API responds with a list, based on our current understanding of the AWS SSO service that API can only ever return zero or one instances. Therefore, for this pull request, we have coded that data source to only ever return the first instance in the response from that API and we return an error if there are more than one instances returned.

https://github.com/hashicorp/terraform-provider-aws/blob/1b750b9ce4e519cfff995df4bbf309d956d53a0e/aws/data_source_aws_sso_instance.go#L53-L55

We could not find a way to create multiple AWS SSO Instances from the AWS console, so we think that AWS just decided to design the [ListInstances](https://docs.aws.amazon.com/singlesignon/latest/APIReference/API_ListInstances.html) API like the [ListAccountAliases](https://docs.aws.amazon.com/IAM/latest/APIReference/API_ListAccountAliases.html) in that they can only ever return zero or one instances, but our assumption could be flawed here. It may be a good idea to get in contact with AWS to see if this assumption is correct. It would be unfortunate if it was possible for there to be more than one sso instance because that would mean it would require users of the `aws_ssoadmin_instance` data source to hardcode their instance arn and identity store id in their terraform as sso instances don't have any other non-id properties (like a name) that we can filter on.

**Comment by anGie44** [comment]
>  It would be unfortunate if it was possible for there to be more than one sso instance because that would mean it would require users of the `aws_ssoadmin_instance` data source to hardcode their instance arn and identity store id in their terraform as sso instances don't have any other non-id properties (like a name) that we can filter on.

Hmm given the API method seems to imply more than 1, we could proactively account for this by making the data source plural i.e. have a field for `instance_arns` and one for `identity_store_ids` . the `data_source_aws_efs_access_points` comes to mind in keeping track of multiple plural fields.

**Review by anGie44 — CHANGES_REQUESTED** [review]
Hi @burck1 @lawdhavmercy -- thanks again for your time/replies! This looks great overall 👍  please find below my initial comments, a couple nitpicks here and there to keep inline with the current conventions we use w/in the CRUD operations as well as comments regarding the provisioning.. I can definitely confirm the need for it even on just a description update as you've kindly provided CLI output earlier. 

I do want to mention for the moment we want to have the policy arguments as separate resources to maintain how we manage these type of resources across the provider (e.g. `organizations_policy_attachment`, `iot_policy` resource etc). I imagine the error we could potentially run into is the `ConflictException` (or maybe even ThrottlingException) when triggering `ProvisionPermissionSet` more than once w/in the same account, though this can be handled with the `resource.Retry` usage, catching the 2 types of errors and returning `resource.RetryableError(err)`. But before moving things out from the `permission_set` resource, let me give it a try locally and confirm back here so you don't lose efforts moving things back and forth from something that is currently functioning.

**Comment by burck1** [comment]
> This looks great overall 👍 please find below my initial comments, a couple nitpicks here and there to keep inline with the current conventions we use w/in the CRUD operations as well as comments regarding the provisioning. I can definitely confirm the need for it even on just a description update as you've kindly provided CLI output earlier.

Perfect! We will review and update.

**Comment by burck1** [comment]
> I do want to mention for the moment we want to have the policy arguments as separate resources to maintain how we manage these type of resources across the provider (e.g. `organizations_policy_attachment`, `iot_policy` resource etc). I imagine the error we could potentially run into is the `ConflictException` (or maybe even `ThrottlingException`) when triggering `ProvisionPermissionSet` more than once w/in the same account, though this can be handled with the `resource.Retry` usage, catching the 2 types of errors and returning `resource.RetryableError(err)`. But before moving things out from the `permission_set` resource, let me give it a try locally and confirm back here so you don't lose efforts moving things back and forth from something that is currently functioning.

One other thought specific to the inline policies that makes it potentially different from the `organizations_policy`, `aws_iot_policy`, or even `aws_iam_role_policy`, permission sets can only have one inline policy.

```sh
# inline policy initially empty
$ aws sso-admin get-inline-policy-for-permission-set --instance-arn "arn:aws:sso:::instance/ssoins-9999999999999999" --permission-set-arn "arn:aws:sso:::permissionSet/ssoins-9999999999999999/ps-1111111111111111"
{
    "InlinePolicy": ""
}

# set inline policy
$ aws sso-admin put-inline-policy-to-permission-set --instance-arn "arn:aws:sso:::instance/ssoins-9999999999999999" --permission-set-arn "arn:aws:sso:::permissionSet/ssoins-9999999999999999/ps-1111111111111111" --inline-policy "{\"Version\":\"2012-10-17\",\"Statement\":[{\"Effect\":\"Allow\",\"Action\":\"s3:ListAllMyBuckets\",\"Resource\":\"*\"}]}"

# get the inline policy
$ aws sso-admin get-inline-policy-for-permission-set --instance-arn "arn:aws:sso:::instance/ssoins-9999999999999999" --permission-set-arn "arn:aws:sso:::permissionSet/ssoins-9999999999999999/ps-1111111111111111"
{
    "InlinePolicy": "{\"Version\":\"2012-10-17\",\"Statement\":[{\"Effect\":\"Allow\",\"Action\":\"s3:ListAllMyBuckets\",\"Resource\":\"*\"}]}"
}

# update inline policy
$ aws sso-admin put-inline-policy-to-permission-set --instance-arn "arn:aws:sso:::instance/ssoins-9999999999999999" --permission-set-arn "arn:aws:sso:::permissionSet/ssoins-9999999999999999/ps-1111111111111111" --inline-policy "{\"Version\":\"2012-10-17\",\"Statement\":[{\"Effect\":\"Allow\",\"Action\":\"s3:ListBucket\",\"Resource\":\"*\"}]}"      

# get the updated inline policy
$ aws sso-admin get-inline-policy-for-permission-set --instance-arn "arn:aws:sso:::instance/ssoins-9999999999999999" --permission-set-arn "arn:aws:sso:::permissionSet/ssoins-9999999999999999/ps-1111111111111111"
{
    "InlinePolicy": "{\"Version\":\"2012-10-17\",\"Statement\":[{\"Effect\":\"Allow\",\"Action\":\"s3:ListBucket\",\"Resource\":\"*\"}]}"
}
```

What this means is if we setup a separate resource, say `aws_ssoadmin_permission_set_inline_policy`, then users would run into a problem if they were to create multiple `aws_ssoadmin_permission_set_inline_policy` resources for a single permission set. So if users were to say have the following:

```hcl
resource "aws_ssoadmin_permission_set" "example" {
  name         = "Example"
  # ..
}

resource "aws_ssoadmin_permission_set_inline_policy" "a" {
  permission_set_arn = aws_ssoadmin_permission_set.example.arn
  policy             = "..."
}

resource "aws_ssoadmin_permission_set_inline_policy" "b" {
  permission_set_arn = aws_ssoadmin_permission_set.example.arn
  policy             = "..."
}
```

The `a` and `b` resources would both override each other when calling [PutInlinePolicyToPermissionSet](https://docs.aws.amazon.com/singlesignon/latest/APIReference/API_PutInlinePolicyToPermissionSet.html), the second resource would always _win_, and therefore the state would always be out of sync for one of them. From code, I don't think there would be an easy way for us to detect the difference between an update to the inline policy or a duplicate resource.

**Comment by burck1** [comment]
> > It would be unfortunate if it was possible for there to be more than one sso instance because that would mean it would require users of the `aws_ssoadmin_instance` data source to hardcode their instance arn and identity store id in their terraform as sso instances don't have any other non-id properties (like a name) that we can filter on.
> 
> Hmm given the API method seems to imply more than 1, we could proactively account for this by making the data source plural i.e. have a field for `instance_arns` and one for `identity_store_ids` . the `data_source_aws_efs_access_points` comes to mind as a reference in keeping track of multiple plural fields.

For the proposal:

```hcl
data "aws_ssoadmin_instances" "selected" {}

locals {
  instance_arn      = data.aws_ssoadmin_instances.selected[0].arn
  identity_store_id = data.aws_ssoadmin_instances.selected[0].identity_store_id
}

data "aws_sso_permission_set" "example" {
  instance_arn = local.instance_arn
  name         = "Example"
}

data "aws_identitystore_group" "example_group" {
  identity_store_id = local.identity_store_id
  display_name      = "ExampleGroup@example.com"
}
```

Some concerns I have since the [ListInstances](https://docs.aws.amazon.com/singlesignon/latest/APIReference/API_ListInstances.html) API doesn't provide a lot of info.

- The API doesn't provide any guarantees about the order of the instances in the response. This means if it is possible to have multiple instances and a user selects the first instance, it's possible for the order to change during the next plan/apply and cause unexpected changes.
- If it is possible to have multiple instances, the docs don't provide any info for if the `InstanceArn` or the `IdentityStoreId` properties are unique or if they can be repeated. I.E. Can we have multiple instances returned from that API where the `InstanceArn` is the same but the `IdentityStoreId` is different. Or maybe vice versa. This means we couldn't simply add an `instance_arn` parameter because there could still be multiple instances in the response with the same `InstanceArn`.

**Comment by anGie44** [comment]
Hi @lawdhavmercy, I think for the time being, given we don't have much about the future number of instances available to a user and given the quota is set to 1 store (https://docs.aws.amazon.com/singlesignon/latest/userguide/limits.html), which makes me thing we're capped at 1 instance atm, let's keep that data source as-is 👍 if in the future, the service quota/console allows for more than 1, the data-source can be adjusted.

File: aws/resource_aws_sso_permission_set.go
Comment (anGie44) [inline aws/resource_aws_sso_permission_set.go:487]: Iooking at how cloudformation doe this, I think provisioning to `ALL` accounts makes sense to make the uniform change, I would just note this in the documentation as it occurs behind the scenes

File: aws/internal/keyvaluetags/sso_tags.go
Comment (anGie44) [inline aws/internal/keyvaluetags/sso_tags.go:12]: let's use the generator available in the provider to get these functions, of course with the exception that we need to account for the `InstanceArn` param .. following along here should be a good start: https://github.com/hashicorp/terraform-provider-aws/blob/master/docs/contributing/contribution-checklists.md#adding-service-to-tag-generating-code

to get the `InstanceArn` param in place, going to the `service_generation_customizations.go` file and then the the func `ServiceTagResourceTypeField(serviceName string)` , there you can add the customization like:
```go
func ServiceTagResourceTypeField(serviceName string) string {
	switch serviceName {
	...
	case "ssm":
		return "ResourceType"
	case "ssoadmin":
		return "InstanceArn"
	default:
		return ""
	}
}
```

File: aws/data_source_aws_sso_instance.go
Comment (burck1) [reply aws/data_source_aws_sso_instance.go:65]: Here the intention was to have an ID which is a combination of the `InstanceArn` and the `IdentityStoreId`. That way if a user were to change their [identity source for AWS SSO](https://docs.aws.amazon.com/singlesignon/latest/userguide/manage-your-identity-source.html), the ID would change. Looking back I don't think this is necessary. Accepting the suggested change.

File: aws/resource_aws_sso_permission_set.go
Comment (burck1) [reply aws/resource_aws_sso_permission_set.go]: resolved

File: aws/resource_aws_sso_permission_set_test.go
Comment (burck1) [reply aws/resource_aws_sso_permission_set_test.go]: resolved

**Comment by jamesgoodhouse** [comment]
hey all! awesome work on this PR. is there any help that's needed to push it across the finish line? I'd be happy to jump in if so

**Review by anGie44 — CHANGES_REQUESTED** [review]
Hi @lawdhavmercy @burck1, thanks again for the pushed up changes! To give you a heads up, since as mentioned earlier we're following a design on our end that separates the permission set from the policy attributes, I will be building on your PR to address needed changes and will comment along the way where these changes occur. Make sure that you have checked the "allow edits from maintainers" box. You will of course get credit for the feature and commits. I'll make any changes needed to bring it into alignment with the design so we can see this feature in a forthcoming release :)

**Comment by anGie44** [comment]
Thanks again for your all your contributions and time @burck1 @lawdhavmercy 🚀 🚀

**Pull request merged**
